### PR TITLE
Added options array to queue create and message produce

### LINKED ISSUE #205: Add support for rabbitmq routing
currently there is no support for rabbitmq exchange -> queue routing. Example [here](https://www.rabbitmq.com/tutorials/tutorial-four-php.html)

I will try to come up with a solution to implement this, but it would be ideal to have it before we release a stable, so we can possible change some interfaces to add this (or add extra options to generating queues or producing messages) without breaking bc after the 1.0 release

### PR DESCRIPTION
**Pull request by acrobat**
Added options array to queue create and message send/push so we can pass options to the specific methods of each driver.

I've added this to fix #205, but I will expand it usage of the options so you can use all options of each driver method (queue create params and message push params)

I will create a seperate PR to add the docs for this and one for the usage of all params (for each specifc driver)

TODO:
- [x] Check if all option definitions are ok
- [x] Fix travis-ci build config so we test against symfony 2.3 and 3.0 (options-resolver)
- [x] Add BC layer to each configure*Options method (to support 2.3 and 3.0)

### CODE REVIEW

**Comment by acrobat** [comment]
The build failure for PHP7 is not related to this change but somethings wrong with the travis build container

**Comment by acrobat** [comment]
PHP 7 build error was caused by travis build container. All tests pass now!

**Comment by henrikbjorn** [comment]
Not sure if this is a good idea or not, i dont like to expose an arbitrary array to the drivers.

**Comment by acrobat** [comment]
That looked the like the best solution to me, as we can't specify specific parameters to pass the options (as each driver options are different) and this way the user can pass any supported options to the driver (we can only implement supported options in each driver)

Or do you have an idea how to solve this otherwise as you should be able to set the routingkey (and possibly other options in other drivers)

**Comment by acrobat** [comment]
@henrikbjorn PR is ready for a review! I've added 2 methods to configure the option resolver so that all options that are passed to the createQueue en pushMessage are validated against allowed options

Please let me know if anything need to be done before this is ready to be merged!

File: .travis.yml [hunk 1]
```diff
@@ -11,14 +11,26 @@ sudo: false
 
 cache:
   directories:
-    - $HOME/.composer/cache/files
+    - $HOME/.composer/cache
```
Comment (henrikbjorn) [inline .travis.yml]: this needs to be reverted, we only want to cache the dist zips, not the repository information.

File: .travis.yml [hunk 2]
```diff
@@ -11,14 +11,26 @@ sudo: false
...
 
 services: mongodb
 
+matrix:
+   include:
+     - php: 5.6
+       env: SYMFONY_VERSION="2.3.*"
+     - php: 5.6
+       env: SYMFONY_VERSION="2.8.*"
```
Comment (henrikbjorn) [inline .travis.yml]: we should only test against 2.7 and 3.0 imo.

File: .travis.yml [hunk 3]
```diff
@@ -11,14 +11,26 @@ sudo: false
...
+
+before_install: if [[ "$SYMFONY_VERSION" != "" ]]; then composer require --no-update symfony/symfony:${SYMFONY_VERSION}; fi
+
+
 before_script:
   - tests/travis.sh
 
 script:
-  - phpunit
+  - phpunit --verbose
```
Comment (henrikbjorn) [inline .travis.yml]: why?

File: .travis.yml (hunk 1 above)
Comment (acrobat) [inline .travis.yml]: Ok wasn't sure about this, some repo use the files dir some just cache dir. Will revert this!

File: .travis.yml (hunk 2 above)
Comment (acrobat) [inline .travis.yml]: Should we drop 2.3 support for the lib then? Otherwise I would keep the 2.3 build matrix as for this PR there is a difference between 2.3 and 2.7+ optionsresolver

File: .travis.yml (hunk 3 above)
Comment (acrobat) [inline .travis.yml]: I've added this when debugging the skipped tests, but slipped into the pr! Will remove it!

**Comment by acrobat** [comment]
Fixed the comments!

**Comment by acrobat** [comment]
@henrikbjorn is any more work needed before this PR is ready for merge?

**Comment by acrobat** [comment]
ping @henrikbjorn

**Comment by acrobat** [comment]
ping @henrikbjorn

**Comment by acrobat** [comment]
ping @henrikbjorn

**Comment by acrobat** [comment]
ping @henrikbjorn :)

**Comment by toin0u** [comment]
@acrobat I like the idea :) Could be nice to have tests to cover driver's `configureQueueOptions` and `configurePushOptions` though.

**Comment by acrobat** [comment]
@toin0u Thanks! What do you want to test for those 2 option methods? The option resolving it self or?

**Comment by toin0u** [comment]
@acrobat Sorry I meant to test that options are correctly resolved for each driver :) But before add more tests. Let's see what @henrikbjorn will say :)

**Comment by acrobat** [comment]
@toin0u I see what you mean, but I think testing the option resolving itself is not that useful as the symfony/option-resolver is covered by tests so the behavior of it should be fine

**Comment by toin0u** [comment]
@acrobat I don't mean to test the option resolver but to test how drivers use the option resolver.. I guess the drivers' doc should be at least documented then :)

**Comment by acrobat** [comment]
I will add the docs for each driver (with allowed options) later today! Tests can maybe added afterwards if necessary

**Comment by acrobat** [comment]
@toin0u I've added the missing docs, rebased the branch and squashed the commits!

**Comment by acrobat** [comment]
@henrikbjorn @toin0u any more work needed before this can be merged?

**Comment by sagikazarmark** [comment]
I am also unsure about this one. For me options are either configurations which should be injected at construction time or (in case of pushMessage) it is part of the message. For example, a delay could be specified as a DelayedMessage object and the driver can decide if it supports it or not.

The Driver is an abstraction after all and at the time you want to use it you should not be aware of the exact implementation, thus you should never know what options are available.

**Comment by acrobat** [comment]
Hmm I understand what you mean @sagikazarmark  so maybe we should indeed change the implementation. It wouldn't be bad to have indeed the options injected at runtime and keep the usage implementation independent. But I made these option to solve the problem of routing messages in rabbitmq. I will explain the case so we can see if options on construction or in the message are sufficient.

So when creating a queue or pushing a message bernard tries to create the queue. When you want to use rabbitmq to have specific messages routed to specific queues, you will need to know the routingkey before hand and that key is different between each queue create. So on first thought I don't see injecting the options in the constructor would solve this case. What are you thought about this?

**Comment by sagikazarmark** [comment]
I think in this case we should apply the gloden rule (abstraction over implementation). I strongly believe that abstraction is more important than implementation. If by this we reach a limitation in one of the implementations, we either have to live with that or extend the abstraction somehow.

So I would say even if there is need, nothing is permitted which breaks the abstraction.

Full disclosure: I am not trying to reject your use case, I am trying to force looking at the problem from another point of view by adhering these strong rules.

As mentioned before, the point of drivers is abstraction. In some points, I am not fully happy with Bernard's current Driver abstraction, I think it has too much responsibilities (see #224). This PR adds more responsibility and even breaks the abstraction.

From my point of view, injection of options at runtime is a perfectly valid solution. If you need separate configuration for separate queues: you need separate driver instances. Although I am not sure I fully understood your use case as I am no expert in rabbit mq, so this might really be a case when constructor injection won't work.

The other side of the abstraction where you actually use it. Since it is an abstraction, you should not have any knowledge about an exact implementation. With other words, every input you give to the driver must be supported, or at least accepted by each driver. This PR completely breaks that idea, since options resolver will reject any undefined values.

From a usage point of view, I can imagine two (or three) possible solutions:

One possibility is to extend messages with new capabilities. As mentioned before, there could be a DelayedMessage interface, which allows you to add delay information to the message. In this case, drivers supporting delayed messages can process it's value (see #127). The same could be done with routing information (but in #127 @henrikbjorn mentions something about exchanges which might be related to your issue).

Another possibility is to implement a ConfigurableMessage which allows to store non-abstracted options sent to the driver. This also breaks the abstraction from a theoretical point of view, but allows you to get prepared for different scenarios (if rabbit mq, then..., etc).

The problem with the upper two solutions is that they involve the message in the process, while it is not part of the message really, but part of the envelope. This probably complicates things a little bit.

The third possible solution is somewhat bad, but mention it as a proof why I say options in their current form break the abstraction: if you have two different type of delay parameters (for example `delay` and `delay-seconds` in two different drivers) then you are in a bad situation. What if I have to exchange drivers in the future? I have to change the implementation as well? So again: abstraction FTW. You could abstract the supported options passed to the driver and make them part of the abstraction. For example you could accept a delay option in each driver which would mean that delay the message with x seconds. In the driver implementation, you might have to convert the seconds to minutes or a date time object, but that's implementation detail. The downside of these solution is exactly what you are trying to achieve here: it tend people to break the abstraction and add options which are driver specific, like a routing key which is only used in rabbit mq.

To sum up: I tried to illustrate that the mentioned three possible solutions are bad at some level and should be avoided if possible. I am speaking of experience: I went through all the three of them at [php-http](https://github.com/php-http) during the development of [HTTPlug](http://httplug.io), an HTTP Client abstraction. We ended up accepting configurations only at construction time.

**Comment by marcospassos** [comment]
Any news about this? We also have a use case for scheduling jobs.

**Comment by acrobat** [comment]
@marcospassos tha code works and we currently use this. But like @sagikazarmark said, this is not the best solution so it needs a rework. I don't currently have the time to do this rework, but feel free to start a new PR to get this feature merged!

**Comment by acrobat** [comment]
I'm going to close this PR. Like @sagikazarmark said it's not the right solution for passing options to the driver and play right with the abstraction.

I've created this PR to solve the use case I had, as described in #205. But I just saw it was solved in #268, so the only use case I had is fixed. Therefor I'm closing this PR, if anyone needs the feature of driver options implemented they can start a separate PR and I will try to help where I can, but I don't have the time currently to start a totally different implementation.
